### PR TITLE
Allow kicking dead players

### LINKED ISSUE #19926: Allow kicking of defeated players
From Discord:

[12:34 PM] Longely: Didnt it get introduced that people who lost got moved to spec this release?
[12:35 PM] Longely: seems like you still cant kick them after they lost.. which is a problem because people stay after they lost and lag the game on purpose to destroy it for everyone else (edited)
[12:44 PM] pchote: i don't think anybody thought of that before now
[12:45 PM] pchote: sounds like a reasonable thing to support
[12:45 PM] pchote: someone should make an issue
[12:50 PM] Longely: oh ok, i thought we talked about it earlier. maybe im confussing it with one of the tons of other trolling methods being used in ffa, maybe the pausing as spec
[12:50 PM] pchote: a bunch of things were fixed to make defeated players behave like spectators, but they are fundamentally still not spectators
[12:50 PM] pchote: and the kick button wasn't thought about as part of the fake spectator fixes

### PR DESCRIPTION
**Pull request by PunkPun**
Closes #19926

With months of contributor inactivity I'm superseding #20083. I've cleaned up the code and yamls.

### CODE REVIEW

File: OpenRA.Game/Server/Server.cs [hunk 1]
```diff
@@ -1323,22 +1329,22 @@ public void StartGame()
...
+					StartTimeUtc = DateTime.UtcNow,
+				};
+
+				// Replay metadata should only include the playable players
+				foreach (var p in worldPlayers)
+					if (p != null)
+						gameInfo.Players.Add(p);
 
+				if (recorder != null)
+				{
```
Comment (dragunoff) [inline OpenRA.Game/Server/Server.cs]: Here `gameInfo` is only needed if `recorder != null` so I think this was fine the way it used to be.

Comment (PunkPun) [reply OpenRA.Game/Server/Server.cs]: GameInfo is used for setting the player states. It is always needed

Comment (dragunoff) [reply OpenRA.Game/Server/Server.cs]: Oh, yes - sorry. I thought it's a local var.

**Review by dragunoff — DISMISSED** [review]
LGTM :+1:

File: OpenRA.Game/Server/Server.cs [hunk 2]
```diff
@@ -1323,22 +1329,22 @@ public void StartGame()
...
 					recorder.Metadata = new ReplayMetadata(gameInfo);
 				}
```
Comment (abcdefg30) [inline OpenRA.Game/Server/Server.cs]:
```suggestion
					recorder.Metadata = new ReplayMetadata(gameInfo);
```

File: OpenRA.Game/Server/Server.cs [hunk 3]
```diff
@@ -1163,6 +1163,12 @@ public Session.Client GetClient(Connection conn)
 			return LobbyInfo.ClientWithIndex(conn.PlayerIndex);
 		}
 
+		/// <Remarks> Does not check if client is admin </Remarks>
+		public bool CanKickClient(Session.Client client)
```
Comment (abcdefg30) [inline OpenRA.Game/Server/Server.cs]: This appears to be only used once, and not even in this class?

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/GameInfoStatsLogic.cs [hunk 4]
```diff
@@ -165,6 +165,24 @@ public GameInfoStatsLogic(Widget widget, ModData modData, World world, OrderMana
 					muteCheckbox.IsVisible = () => !pp.IsBot && client.State != Session.ClientState.Disconnected && pp.ClientIndex != orderManager.LocalClient?.Index;
 					muteCheckbox.GetTooltipText = () => muteCheckbox.IsChecked() ? unmuteTooltip : muteTooltip;
 
+					var kickButton = item.Get<ButtonWidget>("KICK");
+					kickButton.IsVisible = () => Game.IsHost && client.Index != orderManager.LocalClient?.Index && client.State != Session.ClientState.Disconnected && pp.WinState == WinState.Lost && !pp.IsBot;
+					kickButton.OnClick = () =>
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Widgets/Logic/Ingame/GameInfoStatsLogic.cs]: Can we share this logic with the kick button for spectators? (It seems to be identical.)

File: OpenRA.Game/Server/Server.cs (hunk 3 above)
Comment (PunkPun) [reply OpenRA.Game/Server/Server.cs]: Yes, we need access to worldPlayers variable which is supposed to be private on the server

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/GameInfoStatsLogic.cs (hunk 4 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Widgets/Logic/Ingame/GameInfoStatsLogic.cs]: abstracted

File: OpenRA.Game/Server/Server.cs (hunk 3 above)
Comment (abcdefg30) [reply OpenRA.Game/Server/Server.cs]: I see.

File: OpenRA.Game/Server/Server.cs [hunk 5]
```diff
@@ -791,9 +791,12 @@ void OutOfSync(int frame)
 
 			// Make sure the written file is not valid
 			// TODO: storing a serverside replay on desync would be extremely useful
-			recorder.Metadata = null;
+			if (recorder != null)
```
Comment (pchote) [inline OpenRA.Game/Server/Server.cs:794]: Why is this needed?

File: OpenRA.Game/Server/Server.cs [hunk 6]
```diff
@@ -1163,6 +1163,12 @@ public Session.Client GetClient(Connection conn)
 			return LobbyInfo.ClientWithIndex(conn.PlayerIndex);
 		}
 
+		/// <Remarks> Does not check if client is admin </Remarks>
+		public bool CanKickClient(Session.Client client)
+		{
+			return State != ServerState.GameStarted || client.IsObserver || worldPlayers.FirstOrDefault(p => p?.ClientIndex == client.Index)?.Outcome == WinState.Lost;
```
Comment (pchote) [inline OpenRA.Game/Server/Server.cs]: Shouldn't this be `!= WinState.Undefined`?

File: OpenRA.Game/Server/Server.cs (hunk 5 above)
Comment (PunkPun) [reply OpenRA.Game/Server/Server.cs:794]: without metadata the server isn't aware of the state of players

File: OpenRA.Game/Server/Server.cs (hunk 6 above)
Comment (PunkPun) [reply OpenRA.Game/Server/Server.cs]: It could if we want to allow you to kick all players after the game has ended.

File: OpenRA.Game/Server/Server.cs (hunk 5 above)
Comment (PunkPun) [reply OpenRA.Game/Server/Server.cs:794]: Though I guess here specifically, since we now also handle sync orders without the existence of recorder we now need to deal with it being null in more places

File: OpenRA.Game/Server/Server.cs (hunk 6 above)
Comment (PunkPun) [reply OpenRA.Game/Server/Server.cs]: Changed it to undefined

**Comment by tux314159** [comment]
Thanks pun!

File: OpenRA.Game/Server/Server.cs [hunk 7]
```diff
@@ -856,17 +859,14 @@ public void DispatchOrdersToClients(Connection conn, int frame, byte[] data)
 
 		void RecordOrder(int frame, byte[] data, int from)
 		{
-			if (recorder != null)
-			{
-				recorder.ReceiveFrame(from, frame, data);
+			recorder?.ReceiveFrame(from, frame, data);
```
Comment (anvilvapre) [inline OpenRA.Game/Server/Server.cs:862]: this is probably ok, but,
this doesn't change functionality?
i.e. recorders it always set prior if needed?
i.e. data below can safely be handled even if recorder is null?

File: OpenRA.Game/Server/Server.cs [hunk 8]
```diff
@@ -1163,6 +1163,16 @@ public Session.Client GetClient(Connection conn)
...
 		}
 
+		/// <Remarks> Does not check if client is admin </Remarks>
+		public bool CanKickClient(Session.Client client)
+		{
+			if (State != ServerState.GameStarted || client.IsObserver)
+				return true;
+
+			var player = worldPlayers.FirstOrDefault(p => p?.ClientIndex == client.Index);
+			return player != null && player.Outcome != WinState.Undefined;
```
Comment (anvilvapre) [inline OpenRA.Game/Server/Server.cs:1173]: the gui may not offer the option, but it may be good to check if you are no trying to kick yourself

`&& player != Admin|Host && player != self?`

happens often enough the people deadmin themselves.

File: OpenRA.Game/Server/Server.cs (hunk 7 above)
Comment (PunkPun) [reply OpenRA.Game/Server/Server.cs:862]: having the code below run unconditionally is what matters here

File: OpenRA.Game/Server/Server.cs (hunk 8 above)
Comment (PunkPun) [reply OpenRA.Game/Server/Server.cs:1173]: It checks for admin, see the kick code in LobbyCommands.cs

File: OpenRA.Game/Server/Server.cs (hunk 5 above)
Comment (abcdefg30) [reply OpenRA.Game/Server/Server.cs:794]: Why/where does this PR change the cases where `recorder` is null?

Comment (PunkPun) [reply OpenRA.Game/Server/Server.cs:794]: You can see the changelog. It's rather short. 

For the server to know what the state of the game is, it needs to handle sync orders unconditionaly. That is regardless of the existance of recorder. But if we handle unconditionaly, then all the extra code we run needs to be aware of the posibility that recorder is null

Comment (abcdefg30) [reply OpenRA.Game/Server/Server.cs:794]: Ah right, we call `HandleSyncOrder` now.

File: OpenRA.Game/Server/Server.cs (hunk 8 above)
Comment (abcdefg30) [reply OpenRA.Game/Server/Server.cs:1173]: Technically it does only check if the admin executes the command, not if the admin is the client to kick. So an admin could kick themselves.

Comment (PunkPun) [reply OpenRA.Game/Server/Server.cs:1173]: Added the check

File: mods/common/languages/en.ftl [hunk 9]
```diff
@@ -64,8 +64,9 @@ notification-invalid-configuration-command = Invalid configuration command.
 notification-admin-option = Only the host can set that option.
 notification-error-number-teams = Number of teams could not be parsed: { $raw }
 notification-admin-kick = Only the host can kick players.
+notification-kick-self = Host is not allowed to kick himself.
```
Comment (abcdefg30) [inline mods/common/languages/en.ftl]:
```suggestion
notification-kick-self = The host is not allowed to kick themself.
```

File: mods/common/languages/en.ftl [hunk 10]
```diff
@@ -64,8 +64,9 @@ notification-invalid-configuration-command = Invalid configuration command.
 notification-admin-option = Only the host can set that option.
 notification-error-number-teams = Number of teams could not be parsed: { $raw }
 notification-admin-kick = Only the host can kick players.
+notification-kick-self = The host is not allowed to kick himself.
```
Comment (Mailaender) [inline mods/common/languages/en.ftl]: This is still only addressing a male audience.

Comment (PunkPun) [reply mods/common/languages/en.ftl]: It's addressing everyone, regardless of gender

Comment (Mailaender) [reply mods/common/languages/en.ftl]: [The generic masculine in English seems to work differently as for example in German.](https://community.beck.de/2020/12/01/generic-masculine-nouns-and-pronouns-in-english-das-generische-maskulinum) 
> You use themselves instead of 'himself or herself' to refer back to the person who is the subject of a [sentence](https://www.collinsdictionary.com/dictionary/english/sentence) without [saying](https://www.collinsdictionary.com/dictionary/english/saying) whether it is a man or a woman.

Source: [themselves](https://www.collinsdictionary.com/dictionary/english/themselves) 3rd definition in Collins Dictionary

Comment (PunkPun) [reply mods/common/languages/en.ftl]: sure, this is better than `themself` suggestion

Comment (Mailaender) [inline mods/common/languages/en.ftl]:
```suggestion
notification-kick-self = Hosts are not allowed to kick themselves.
```
maybe?

Comment (PunkPun) [reply mods/common/languages/en.ftl]: fwiw english has generic masculine, but often english users use linguistic strategies to make it non-gender specific or invent new words

Comment (abcdefg30) [reply mods/common/languages/en.ftl]: Bruh.

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ca86f742643fbb3ec7aa278d444387f4387711c0...c0acc64efc407d891e473a3f9017f8d82dbe7d41)
